### PR TITLE
A proposal for the assign operator overload

### PR DESCRIPTION
**Pull request by asodja**
This is a proposal on providing an assign operator overload that would provide DSL for assigning values to mutable container objects.

Please, use [KEEP-309](https://github.com/Kotlin/KEEP/issues/309) issue for discussion of this proposal and post text and editorial changes into this PR.

### CODE REVIEW

File: proposals/assign-operator-overload.md [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+myTask.input2 = calculatedProperty
+myTask.output = File("input-file.txt")
+```
+
+Or as a DSL
+
+```kotlin
+myTask.apply {
+    input1 = File("input-file.txt")
+    Input2 = calculatedProperty
```
Comment (alllex) [inline proposals/assign-operator-overload.md:126]: `Input2` is accidentally capitalized here.
There are also a couple more places in the document where this typo has leaked.

**Comment by JakeWharton** [comment]
Also, in general, opinions on the proposal should go in the associated issue (which hopefully is coming) rather than on the PR where its comments should be reserved for improving only the language / content / syntax of the proposal.

File: proposals/assign-operator-overload.md (hunk 1 above)
Comment (asodja) [reply proposals/assign-operator-overload.md]: Thank you, should be fixed now.

File: proposals/assign-operator-overload.md [hunk 2]
```diff
@@ -0,0 +1,440 @@
+# Assign operator overload
+
+* **Type**: Design proposal
+* **Authors**: Anže Sodja, Lóránt Pintér, Stefan Wolf, Roman Elizarov
+* **Status**: In review
+* **Discussion**: // TODO add KEEP issue
```
Comment (elizarov) [inline proposals/assign-operator-overload.md:6]: Please, link to https://github.com/Kotlin/KEEP/issues/309 here.

File: proposals/assign-operator-overload.md [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+```kotlin
+val list = mutableListOf<Int>()
+list += 1 // operator plusAssign (1)
+list += listOf(2, 3) // operator plusAssign (2)
+list[0] = 4 // operator set
+```
+
+Notice that the `list` is defined as a read-only (`val`) property, but contains a reference to a mutable container of type `MutableList` which is modified with various operators using variants of assign (`=`) operator.
+
+These operators are not hardcoded in Kotlin to operate on lists of arrays, for that matter, but are implemented in the language via _operator conventions_. The language is designed to recognize those conventions and resolve the corresponding operators (in the above example we use operators `plusAssign` and `set`), while the actual support for the specific containers is provided in the standard library. Other libraries are free to define those operators on their own containers to suit the intended usage of those containers.
```
Comment (elizarov) [inline proposals/assign-operator-overload.md:28]: `lists of arrays` -> `lists on arrays`

Comment (asodja) [reply proposals/assign-operator-overload.md]: Fixed

File: proposals/assign-operator-overload.md (hunk 2 above)
Comment (asodja) [reply proposals/assign-operator-overload.md]: Done

**Comment by elizarov** [comment]
Since we've rejected the idea of adding the corresponding convention, as described in the KEEP, to the language directly, but are providing an extension point that allows the similar implementation as a compiler plugin to be used in Gradle, I'm closing this PR and issue.

The documentation for the feature shall change (as only a specific Gradle DSL behavior needs to be documented) and shall be hosted elsewhere. The documentation of the extension points itself will be performed in the future during compiler API stabilization.

**Pull request closed without merging**